SoftLayer/managers/hardware.py

```diff
@@ -780,6 +782,146 @@ def get_hardware_item_prices(self, location):
 
+    def upgrade(self, instance_id, memory=None,
+                nic_speed=None, drive_controller=None,
+                public_bandwidth=None, test=False):
+        """Upgrades a hardware server instance.
+
+        :param int instance_id: Instance id of the hardware server to be upgraded.
+        :param int memory: Memory size.
+        :param string nic_speed: Network Port Speed data.
+        :param string drive_controller: Drive Controller data.
+        :param int public_bandwidth: Public keyName data.
+        :param bool test: Test option to verify the request.
+
+        :returns: bool
+        """
+        upgrade_prices = self._get_upgrade_prices(instance_id)
+        prices = []
+        data = {}
+
+        if memory:
+            data['memory'] = memory
+        if nic_speed:
+            data['nic_speed'] = nic_speed
+        if drive_controller:
+            data['disk_controller'] = drive_controller
+        if public_bandwidth:
+            data['bandwidth'] = public_bandwidth
+
+        server_response = self.get_instance(instance_id)
+        package_id = server_response['billingItem']['package']['id']
+
+        maintenance_window = datetime.datetime.now(utils.UTC())
+        order = {
+            'complexType': 'SoftLayer_Container_Product_Order_Hardware_Server_Upgrade',
+            'properties': [{
+                'name': 'MAINTENANCE_WINDOW',
+                'value': maintenance_window.strftime("%Y-%m-%d %H:%M:%S%z")
+            }],
+            'hardware': [{'id': int(instance_id)}],
+            'packageId': package_id
+        }
+
+        for option, value in data.items():
+            price_id = self._get_prices_for_upgrade_option(upgrade_prices, option, value)
+            if not price_id:
+                # Every option provided is expected to have a price
+                raise exceptions.SoftLayerError(
+                    "Unable to find %s option with value %s" % (option, value))
+
+            prices.append({'id': price_id})
+
+            order['prices'] = prices
+
+        if prices:
+            if test:
+                self.client['Product_Order'].verifyOrder(order)
+            else:
+                self.client['Product_Order'].placeOrder(order)
+            return True
+        return False
+
+    @retry(logger=LOGGER)
+    def get_instance(self, instance_id):
+        """Get details about a hardware server instance.
+
+        :param int instance_id: the instance ID
+        :returns: A dictionary containing a large amount of information about
+                  the specified instance.
+        """
+        mask = [
+            'billingItem[id,package[id,keyName]]'
+        ]
+        mask = "mask[%s]" % ','.join(mask)
+
+        return self.hardware.getObject(id=instance_id, mask=mask)
+
+    def _get_upgrade_prices(self, instance_id, include_downgrade_options=True):
+        """Following Method gets all the price ids related to upgrading a VS.
```

Comment:
VS is for virtual server meaning,  should be HS for hardware server or just hardware I think.

Reply:
Done.